<a href="https://colab.research.google.com/github/shriyashrestha525/Credit-card-fraud-detection/blob/main/Credit_card_fraud_detection.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from google.colab import drive
drive.mount('/content/mydrive/')

Mounted at /content/mydrive/


In [ ]:
path = '/content/mydrive/MyDrive/ML/creditcard.csv'

In [ ]:
import pandas as pd

In [ ]:
data = pd.read_csv(path)

In [ ]:
data.sample(10)

Time        V1        V2        V3        V4        V5        V6  \
146674   87819.0 -1.844327  1.732348 -0.232243 -2.820888  0.284269 -1.067582   
38101    39247.0  1.057802 -0.088960  1.303887  1.296587 -0.695119  0.628143   
199066  132794.0 -0.426472  1.194171 -0.308636 -0.585369  0.253760 -1.316094   
268937  163455.0 -0.294946  1.173798 -1.080458 -1.212467  1.408501 -1.263446   
118277   75020.0  1.112890 -0.076149  1.099711  1.148249 -0.544157  0.721637   
35          26.0 -0.535388  0.865268  1.351076  0.147575  0.433680  0.086983   
110479   71839.0  1.109198  0.014414  1.287071  1.282978 -0.683065  0.293316   
267899  162975.0 -2.265041 -7.794125 -3.138368  0.886906 -2.903928  0.345232   
8418     11285.0 -1.606425 -0.062862  1.147004 -0.553833  0.103641 -0.669282   
117201   74582.0  0.947823  0.097033  0.874389  2.649682 -0.426733  0.147121   

              V7        V8        V9  ...       V21       V22       V23  \
146674  0.795035  0.417976  0.465274  ... -0.322154 -0.785305 -0.073623   
38101  -0.709913  0.360400  0.510866  ...  0.021637  0.273665  0.008122   
199066  0.723721  0.008653 -0.195382  ...  0.403992  1.043020 -0.198399   
268937  1.693430 -0.410966 -0.238880  ...  0.201227  0.823334 -0.205698   
118277 -0.762190  0.384005  0.497268  ...  0.176597  0.688438 -0.106537   
35      0.693039  0.179742 -0.285642  ...  0.049526  0.206537 -0.187108   
110479 -0.546861  0.153060  0.688921  ... -0.019118  0.249646  0.008468   
267899  2.206327 -0.541956  1.126671  ...  1.544102 -0.481910 -1.850035   
8418   -0.118381  0.516329  1.595968  ... -0.029997 -0.041913  0.052248   
117201 -0.059238  0.144236 -0.488208  ... -0.191964 -0.690356  0.003657   

             V24       V25       V26       V27       V28   Amount  Class  
146674 -0.410176  0.127725  0.398088  0.518673  0.331629     1.00      0  
38101   0.024155  0.316278 -0.381896  0.085572  0.025948    12.00      0  
199066  0.086947  0.028829 -0.210653 -0.342058 -0.025270     1.00      0  
268937  0.631750 -0.197086  0.068848  0.495576  0.350313    20.00      0  
118277 -0.288903  0.444819 -0.208873  0.079995  0.017941     1.00      0  
35      0.000753  0.098117 -0.553471 -0.078306  0.025427     1.77      0  
110479  0.107868  0.385894 -0.383452  0.097525  0.037741     9.99      0  
267899 -0.315174 -1.135256  0.478677 -0.465283  0.320237  2235.00      0  
8418    0.099178 -0.232648  0.883525  0.137837 -0.012557    64.94      0  
117201  0.290739  0.300867 -0.204559 -0.013311  0.030151    89.00      0  

[10 rows x 31 columns]

In [ ]:
data.shape

(284807, 31)

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [ ]:
data.duplicated().sum()

1081

In [ ]:
duplicates = data.duplicated()
data = data.drop_duplicates()

In [ ]:
data.duplicated().sum()

0

In [ ]:
data.corr()['Class']

Time     -0.012359
V1       -0.094486
V2        0.084624
V3       -0.182322
V4        0.129326
V5       -0.087812
V6       -0.043915
V7       -0.172347
V8        0.033068
V9       -0.094021
V10      -0.206971
V11       0.149067
V12      -0.250711
V13      -0.003897
V14      -0.293375
V15      -0.003300
V16      -0.187186
V17      -0.313498
V18      -0.105340
V19       0.033631
V20       0.021486
V21       0.026357
V22       0.004887
V23      -0.006333
V24      -0.007210
V25       0.003202
V26       0.004265
V27       0.021892
V28       0.009682
Amount    0.005777
Class     1.000000
Name: Class, dtype: float64

In [ ]:
data['Class'].value_counts()

0    283253
1       473
Name: Class, dtype: int64

This data is highly unbalanced as there are 283253 normal transaction and no. of fraud transaction is only 473. If we give this type of data to our model the model predicts every transaction as legit transaction.

**separate fraud and legit transactions**

In [ ]:
legit = data[data.Class == 0]
fraud = data[data.Class == 1]

In [ ]:
print(fraud.sample(5))

            Time        V1        V2        V3        V4        V5        V6  \
222419  142961.0  0.457845  1.373769 -0.488926  2.805351  1.777386  0.100492   
55401    46925.0 -0.481531  1.059542  0.647117  0.905586  0.819368 -0.091184   
30384    35899.0 -2.857170  4.045601 -4.197299  5.487199 -3.070776 -1.422686   
68633    53076.0  1.296231  0.417447  0.193963  0.901644  0.130531 -0.371634   
220725  142280.0 -1.169203  1.863414 -2.515135  5.463681 -0.297971  1.364918   

              V7        V8        V9  ...       V21       V22       V23  \
222419  1.295016 -0.135857 -1.695822  ...  0.105593  0.371014  0.051105   
55401   0.504135  0.161064 -0.765054  ...  0.070633  0.192491 -0.174659   
30384  -5.651314  2.019657 -5.015491  ...  1.080323 -0.561384  0.102678   
68633   0.158126 -0.202669 -0.079512  ... -0.112114 -0.220002 -0.121022   
220725  0.759219 -0.118861 -2.293921  ... -0.393090 -0.708692  0.471309   

             V24       V25       V26       V27       V28  Amount  Cl

In [ ]:
print(legit.shape)
print(fraud.shape)

(283253, 31)
(473, 31)


In [ ]:
legit.Amount.describe()

count    283253.000000
mean         88.413575
std         250.379023
min           0.000000
25%           5.670000
50%          22.000000
75%          77.460000
max       25691.160000
Name: Amount, dtype: float64

In [ ]:
fraud.Amount.describe()

count     473.000000
mean      123.871860
std       260.211041
min         0.000000
25%         1.000000
50%         9.820000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

In [ ]:
data.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94835.058093  0.013439 -0.009829  0.012853 -0.010440  0.006769   
1      80450.513742 -4.498280  3.405965 -6.729599  4.472591 -2.957197   

             V6        V7        V8        V9  ...       V20      V21  \
Class                                          ...                      
0      0.001251  0.010447 -0.002448  0.002613  ... -0.000489 -0.00115   
1     -1.432518 -5.175912  0.953255 -2.522124  ...  0.405043  0.46655   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.000160  0.000360  0.000393 -0.000301  0.000065  0.001409  0.000418   
1      0.086639 -0.096464 -0.106643  0.040615  0.050456  0.213774  0.078270   

           Amount  
Class              
0       88.413575  
1      123.871860  

[2 rows x 30 columns]

**Under Sampling**
In under sampling a sample dataset is created using original dataset with similar distribution between normal transaction and fraud transaction.

In [ ]:
legit_sample = legit.sample(473)

In [ ]:
legit_sample.shape

(473, 31)

concatenate legit sample and fraud sample together

In [ ]:
new_data = pd.concat([legit_sample, fraud], axis = 0)

In [ ]:
new_data.sample(5)

Time        V1        V2        V3        V4        V5        V6  \
154622  102259.0  0.040188  1.016963  0.448995 -0.395383  0.572248 -0.982367   
271488  164596.0  2.203223 -0.034768 -2.300040 -0.791571  0.640565 -1.437882   
67345    52485.0  0.836893 -2.503014  0.860684 -0.641784 -2.325386  0.485854   
52584    45541.0 -1.519244  2.308492 -1.503599  2.064101 -1.000845 -1.016897   
129547   79129.0 -1.322459  0.974949  0.690552  1.018067 -0.032631 -1.028084   

              V7        V8        V9  ...       V21       V22       V23  \
154622  0.940980 -0.296573  1.146796  ... -0.398798 -0.683695  0.077822   
271488  0.685265 -0.586621  0.320463  ...  0.031328  0.264826 -0.041951   
67345  -1.283646  0.088458 -0.434920  ... -0.306749 -0.752569 -0.281178   
52584  -2.059731 -0.275166 -1.562206  ...  1.307871  0.102826 -0.017746   
129547  0.255402  0.215171 -0.477704  ...  0.138019  0.398861 -0.383371   

             V24       V25       V26       V27       V28  Amount  Class  
154622 -0.064191 -0.507182  0.102805  0.228060  0.091689    1.79      0  
271488 -0.604057  0.273259  1.022998 -0.135517 -0.094419    3.03      0  
67345  -0.089972  0.315882 -0.213726  0.037347  0.079217  330.00      0  
52584   0.149696 -0.096602 -0.369115 -0.019244 -0.208319    1.00      1  
129547  0.424144 -0.268090 -0.340735  0.173964 -0.022504   28.37      0  

[5 rows x 31 columns]

In [ ]:
new_data['Class'].value_counts()

0    473
1    473
Name: Class, dtype: int64

In [ ]:
new_data.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      92396.723044  0.018209 -0.035065  0.117969 -0.070582 -0.088358   
1      80450.513742 -4.498280  3.405965 -6.729599  4.472591 -2.957197   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0     -0.042214  0.028546 -0.050792  0.010100  ...  0.010196  0.053464   
1     -1.432518 -5.175912  0.953255 -2.522124  ...  0.405043  0.466550   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0      0.019213  0.003082  0.013753  0.024778  0.004114 -0.007015 -0.007461   
1      0.086639 -0.096464 -0.106643  0.040615  0.050456  0.213774  0.078270   

           Amount  
Class              
0       94.321755  
1      123.871860  

[2 rows x 30 columns]

Here groupby function shows that the mean value before equalizing the data and after are same so the properties are saved.

**Split data into feature and class**

In [ ]:
x = new_data.drop(['Class'], axis = 'columns')
y = new_data['Class']

In [ ]:
print(x.shape)

(946, 30)


In [ ]:
print(y.shape)

(946,)


In [ ]:
print(x)

            Time        V1        V2        V3        V4        V5        V6  \
221044  142410.0 -1.105383 -0.238779  2.163768 -2.219586  0.869433  1.718298   
19709    30489.0  1.238449  0.280031  0.182916  0.500719 -0.171931 -0.563423   
119553   75491.0 -1.017887  1.243649  1.449462  0.841787  0.179313 -0.757522   
54983    46717.0 -1.467323  1.139328 -0.010334 -1.937085  0.426090 -0.452434   
196202  131405.0  2.177094 -1.742137 -0.252951 -1.755092 -1.526178  0.542347   
...          ...       ...       ...       ...       ...       ...       ...   
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  .

In [ ]:
print(y)

221044    0
19709     0
119553    0
54983     0
196202    0
         ..
279863    1
280143    1
280149    1
281144    1
281674    1
Name: Class, Length: 946, dtype: int64


**Split data into train and test**

In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.2, stratify = y, random_state = 2)

**Training the model**

In [ ]:
from sklearn.ensemble import GradientBoostingClassifier
model = GradientBoostingClassifier()

Training with the data

In [ ]:
model.fit(x_train, y_train)

GradientBoostingClassifier()

In [ ]:
y_pred = model.predict(x_test)
print(y_pred)

[0 1 0 1 1 0 1 0 0 0 0 0 1 1 1 0 0 0 0 1 0 1 1 1 0 0 1 0 0 0 1 0 0 0 0 1 1
 1 0 0 0 0 0 1 0 0 0 1 0 1 0 1 1 1 0 0 1 1 0 1 0 1 0 0 0 1 1 1 0 0 0 0 1 0
 0 0 0 0 1 0 0 1 1 1 0 0 1 1 1 0 1 1 0 1 0 0 0 0 1 1 0 1 0 0 0 0 1 0 0 1 0
 0 1 1 0 1 0 1 0 1 1 1 1 1 1 1 1 1 0 0 1 0 1 1 0 1 1 0 0 0 0 0 0 1 0 0 0 0
 0 1 1 0 1 1 0 1 0 0 1 1 1 0 0 1 0 0 0 0 1 1 0 0 1 0 1 0 1 1 0 1 1 1 1 1 1
 1 1 0 0 0]


**Accuracy testing**

In [ ]:
from sklearn.metrics import accuracy_score

In [ ]:
accuracy = accuracy_score(y_pred, y_test)
print("The accuracy of model is:",accuracy)

The accuracy of model is: 0.9368421052631579


**Classification report**

In [ ]:
from sklearn.metrics import classification_report
report = classification_report(y_test, y_pred)
print(report)

              precision    recall  f1-score   support

           0       0.90      0.98      0.94        95
           1       0.98      0.89      0.93        95

    accuracy                           0.94       190
   macro avg       0.94      0.94      0.94       190
weighted avg       0.94      0.94      0.94       190



**confusion matrix**

In [ ]:
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_pred, y_test)
print(cm)

[[93 10]
 [ 2 85]]


**Hyperparameter Tuning**

In [ ]:
from sklearn.model_selection import GridSearchCV

In [ ]:
param_grid = {
    'n_estimators': [50, 100, 200],
    'learning_rate': [0.01, 0.1, 0.2],
    'max_depth': [3, 4, 5],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}


In [ ]:
grid_search = GridSearchCV(model, param_grid, cv = 5)

In [ ]:
grid_search.fit(x,y)

GridSearchCV(cv=5, estimator=GradientBoostingClassifier(),
             param_grid={'learning_rate': [0.01, 0.1, 0.2],
                         'max_depth': [3, 4, 5], 'min_samples_leaf': [1, 2, 4],
                         'min_samples_split': [2, 5, 10],
                         'n_estimators': [50, 100, 200]})

In [ ]:
#best parameters
best_param = grid_search.best_params_
print(best_param)
#highest accuracy
high_accuracy = grid_search.best_score_
print(high_accuracy)

{'learning_rate': 0.2, 'max_depth': 5, 'min_samples_leaf': 4, 'min_samples_split': 2, 'n_estimators': 200}
0.9397270955165691
